[rocprof-compute] Improve standalone roofline plot generation

jamessiddeley-amd opened the pull request:
## Motivation
Improved standalone roofline visualization by fixing overlapping labels and adding detailed kernel performance information. The new three-panel layout displays: (1) roofline plot with clean legend, (2) plot points table with AI values and memory/compute bound status, and (3) full kernel names table.

### Screenshots
After:
`rocprof-compute profile --name occupancy -VVV  --device 0 -- tests/occupancy -n 1048576 -b 256`
<img width="796" height="789" alt="image" src="https://github.com/user-attachments/assets/5ab80954-a852-47f2-b89f-387cf8c330fe" />

Gui Analysis After: **Slightly outdated image - see comments**
`rocprof-compute analyze --gui -VVV -p workloads/wgemm_legends/MI350/`
<img width="1893" height="903" alt="image" src="https://github.com/user-attachments/assets/440731b2-a0b4-4e28-bcab-b2f9036c6de6" />

## Technical Details

- Moved bandwidth/peak performance values from plot area to legend to remove overlap
- Added a three-subplot layout using plotly `make_subplots` with dynamic height calculation, based on kernels filtered
- Added plot points table showing symbol, cache level, AI, performance, and bound status for each kernel
- Added kernel names table integrated directly into PDF plot generation. This negates the use for `--kernel-names` arg, which has been removed from roofline logic)
- Added `_determine_kernel_bound_status()` to classify kernels as memory or compute bound **per cache level**
- Updated GUI integration to dynamically filter tables when kernels are selected, as well as when cache levels or bandwidth/ceiling lines are clicked on (as it was before).
- Updated logic for plot generation in `roofline_calc` to properly draw line of peak ceilings in plotly to the **end** of it's plot, where before it ended at a max constant of 10000. 

## Test Plan
- Tested standalone mode with multiple datatypes (FP32, FP64, FP4 + FP6, etc)
- Verified GUI filtering updates all three subplots correctly
- Tested with long kernel names (200+ characters) to verify text wrapping (wgemm)
- Ran ctest suite

## Test Result
Ctests tests passed. PDFs generate correctly with proper layout, legend values display without overlap, tables update dynamically with filtering, and bound status classification is accurate.

## Submission Checklist
- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

jamessiddeley-amd committed:
roofline improvements

jamessiddeley-amd committed:
ruff formatting

jamessiddeley-amd committed:
Update roofline.py function descriptions

jamessiddeley-amd committed:
Update height calculation

jamessiddeley-amd committed:
Add back cache level filtering in gui_analysis

jamessiddeley-amd committed:
Update roofline_calc.py to take in ai_data for ceiling length calc

jamessiddeley-amd committed:
format roofline.py

jamessiddeley-amd committed:
update roof test cases

jamessiddeley-amd committed:
update labels and test cases

jamessiddeley-amd committed:
Merge branch 'develop' of github.com:ROCm/rocm-systems into users/jamessiddeley-amd/standalone-plot

benrichard-amd commented:
* Why in the `--gui` case there is the `ai_l1, ai_2, ai_hbm` to the right of the plot, but not in the PDF case?
<img width="215" height="108" alt="image" src="https://github.com/user-attachments/assets/63d0ee5a-6e3e-4710-b384-6460490f211b" />

I actually prefer without. Much cleaner.

*  since we have a `Cache Level` column we don't even need to use the `ai_l` naming, which is confusing anyway. We can use `L1`, 'L2`, and 'HBM`.

* "AI" and "Performance" are redundant. Consider using FLOPS/byte and GFLOPS for those column headers
* Consider not using scientific notation for FLOPS/byte and GFLOPS.  FLOPS/byte is usually very small and you will only get very large values in artificial cases. And GFLOPS will be in the millions at most.

jamessiddeley-amd commented:
> * Why in the `--gui` case there is the `ai_l1, ai_2, ai_hbm` to the right of the plot, but not in the PDF case?
> I actually prefer without. Much cleaner.

The reason I added this only in webgui was because for the old webgui analysis, toggling cache levels required clicking on the respective cache level in the plotly 'plot' legend. This was not immediately possible through interaction with the _new_ plot point 'subplot' table. Plotly's built-in legend toggling works within the same plot, but cross-subplot interactions need a **lot** of custom clientside JavaScript callbacks which would require much more time to implement. Instead (at least for now, definitely may come back to it later), as we were doing before, we simply use plotly's built in interactivity. So essentially the only reason we have this is for interactability in webgui (it doesnt generate in standalone)

> * since we have a `Cache Level` column we don't even need to use the `ai_l` naming, which is confusing anyway. We can use `L1`, 'L2`, and 'HBM`.
> * "AI" and "Performance" are redundant. Consider using FLOPS/byte and GFLOPS for those column headers
> * Consider not using scientific notation for FLOPS/byte and GFLOPS.  FLOPS/byte is usually very small and you will only get very large values in artificial cases. And GFLOPS will be in the millions at most.

Good points, thanks for the clarification. I'll add those now

jamessiddeley-amd committed:
update roofline legend plot table

jamessiddeley-amd committed:
fix pdf generate cutoff

jamessiddeley-amd commented:
Updated pdf:
- changed to standard notation from scientific
- changed plot point table headers
- changed instances of `ai_*` to just `*.upper()` (e.g. `ai_hbm` -> `HBM` in plot and legends and plot table
- fixed up pdf height generation when writing to pdf to prevent random added blank pages

`rocprof-compute profile -VVV --name wgemm --roof-only --roofline-data-type FP32 FP64 --device 0 -- ../rocWMMA/build_gfx942/test/gemm/gemm_PGR1_LB2_MP0_MB_CP_ad_hoc-bench`

<img width="989" height="1134" alt="image" src="https://github.com/user-attachments/assets/7cdcd804-2a61-4832-bdd9-4aadeee57b7d" />

cfallows-amd reviewed (approved):
Right now this PR enables --kernel-names by default- next step would be cleanup task to remove all references to it.

feizheng10 commented:
One minor comment:
I never see anyone else using term like "L1 Compute Bound" while I do understand the meaning here.
(1) "L1" here duplicated the info in column "Cache Level", (2) It is a bit confused when people see it at the 1st eye.
I think the meaning we want to say here is "This kernel is compute bound on L1 Cache Level", right?
So it seems the clearer solution would be: moving column "Cache Level" to the end, and removing the cache-prefix in column "Status". Thought?

jamessiddeley-amd committed:
reordered plot point table headers + removed cache from status

jamessiddeley-amd commented:
> (1) "L1" here duplicated the info in column "Cache Level", (2) It is a bit confused when people see it at the 1st eye.

This is fair. I assumed cache Level should be the very first column after the symbol, since it's the main way the plot points are categorized, followed by its specific attributes (coordinates and status). I have moved cache level to the end and removed string prepending for the cache level in the 'status' message (so now it either only says `Memory Bound`, `Compute Bound`, or `Unknown`. What do you think?

<img width="796" height="789" alt="image" src="https://github.com/user-attachments/assets/24629a06-1c61-420b-9ca7-0ba29ef23ff3" />

PR description has also been updated with this image

jamessiddeley-amd committed:
format ruff

feizheng10 commented:
I totally agree with you "since it's the main way the plot points are categorized". Showing it at the last column doesn't mean we are not following that:)
So you could tell "Think logic" != "Showing logic" in frontend design in this case.

vedithal-amd commented on projects/rocprofiler-compute/tests/test_profile_general.py:
```diff
@@ -156,7 +156,7 @@
 ROOF_ONLY_FILES = sorted([
     "empirRoof_gpu-0_FP32.pdf",
-    "kernelName_legend.pdf",
+    # "kernelName_legend.pdf",
```
Remove unnecessary comment

vedithal-amd reviewed (approved):
minor: Remove unnecessary comment

jamessiddeley-amd committed:
remove unnecessary comment

JeniferC99 commented:
/AzurePipelines run rocm-ci-caller

cfallows-amd committed:
Merge branch 'develop' into users/jamessiddeley-amd/standalone-plot
